src/libcore/sync/atomic.rs

```diff
@@ -405,6 +404,7 @@ impl AtomicBool {
     #[inline]
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn load(&self, order: Ordering) -> bool {
+        // SAFETY: data races are prevented by atomic intrinsics
         unsafe { atomic_load(self.v.get(), order) != 0 }
```

Comment:
We could change the `atomic_*` intrinsics to take references instead of pointers and make them safe. Afaict the only reason they are unsafe is because they take raw pointers

Reply:
There are other reasons. At minimum, atomic operation require natural alignment (align=size) which is sometimes more than the "ABI alignment" required by references (e.g., for u64 on some 32 bit targets).

Reply:
There's also https://github.com/rust-lang/rust/issues/55005, which basically requires the use of raw pointers for some kinds of concurrent code (the kind that synchronizes to handle deallocation).

Reply:
Heh, ok. so alignment and dangliness are what should be mentioned here, not data races

Reply:
Data races are also relevant: this must not race with a non-atomic access to the same memory.

Reply:
So what would be the best thing to say here? I'm not sure what kind of guarantees we can make about `UnsafeCell`.

Reply:
I created another pull request with a smaller number of changes (including `core::sync::atomic`). I'm going to close this request, so it might be better to finish this discussion there: https://github.com/rust-lang/rust/pull/66564